### PR TITLE
Support diktat - Kotlin code style analyzer 

### PR DESCRIPTION
**Pull request by Cheshiriks**
Diktat is a strict coding standard for Kotlin and a custom set of rules for detecting code smells, code style issues and bugs.

https://github.com/cqfn/diKTat/blob/master/README.md

* add DiktatStep
* add tests for diktat

### CODE REVIEW

File: lib/src/main/java/com/diffplug/spotless/kotlin/DiktatStep.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+				DiktatStep.State::createFormat);
+	}
+
+	static final class State implements Serializable {
+
+		private static final long serialVersionUID = 1L;
+
+		/** Are the files being linted Kotlin script files. */
+		private final boolean isScript;
+		private final String configPath;
```
Comment (nedtwigg) [inline lib/src/main/java/com/diffplug/spotless/kotlin/DiktatStep.java:78]: `configPath` should be `FileSignature`, create with `FileSignature.signAsList`, access path with `FileSignature.getOnlyFile()`

The reason is that if this config file changes, the formatting will change too. The `State` class needs to capture every aspect of the formatter or else up-to-date and buildcache behavior will be broken.

**Review by nedtwigg — CHANGES_REQUESTED** [review]
Thanks, great PR! Few minor changes required before merging.

**Comment by nedtwigg** [comment]
- Oh! Also all three CHANGES.md

**Comment by Cheshiriks** [comment]
> Thanks, great PR! Few minor changes required before merging.

Thanks! Added changes to the CHANGES.md files and made the necessary edits.

**Comment by akuleshov7** [comment]
I guess it should be linked with #409

File: plugin-maven/README.md [hunk 2]
```diff
@@ -311,6 +312,19 @@ Groovy-Eclipse formatting errors/warnings lead per default to a build failure. T
...
+<a name="applying-diktat-to-kotlin-files"></a>
+
+### diktat
+
+[homepage](https://github.com/cqfn/diKTat). [changelog](https://github.com/cqfn/diKTat/releases). [code](https://github.com/diffplug/spotless/blob/main/plugin-maven/src/main/java/com/diffplug/spotless/maven/kotlin/Diktat.java). You can provide configuration path manually as `configPath`.
+
+```xml
+<diktat>
+  <version>0.4.0</version> <!-- optional -->
+  <configPath>"full/path/to/diktat-analysis.yml"</configPath> <!-- optional, configuration file path -->
```
Comment (nedtwigg) [inline plugin-maven/README.md:324]: These quotes seem like a mistake. Is that correct?

**Review by nedtwigg — CHANGES_REQUESTED** [review]
Thanks very much, great PR! One minor question left. Don't worry about changing it, just answer the question and I can change it (I've got a few other tiny nits which are easier if I just do them).

Btw, I find it easier to review PRs if you push up new commits, rather than force-pushing over old-ones. We can always squash down if things get messy, but we can't unsquash :)

File: plugin-maven/README.md (hunk 2 above)
Comment (Cheshiriks) [reply plugin-maven/README.md:324]: Yes. These are version and config file settings, but if they are not specified, the defaults will be used.

**Comment by nedtwigg** [comment]
Thanks for a great PR @Cheshiriks! Published in `plugin-gradle 5.10.0` and `plugin-maven 2.8.0`.

**Comment by nedtwigg** [comment]
- [ ] Btw, we should definitely add links to the [diKTat docs](https://github.com/cqfn/diKTat) for [gradle](https://github.com/diffplug/spotless/tree/main/plugin-gradle#diktat) and [maven](https://github.com/diffplug/spotless/tree/main/plugin-maven#diktat).

**Comment by petertrr** [comment]
Hey @nedtwigg, we have added description of spotless integration in diktat readme in https://github.com/cqfn/diKTat/pull/762, you can check it out [here](https://github.com/cqfn/diKTat#run-with-spotless) 👍

**Comment by nedtwigg** [comment]
Great, thanks! It might be worth linking directly to the [gradle](https://github.com/diffplug/spotless/tree/main/plugin-gradle#diktat) and [maven](https://github.com/diffplug/spotless/tree/main/plugin-maven#diktat) pages rather than the Spotless root docs, but it looks great as-is. Spotless is glad to have another formatter, hopefully the formatter is glad to have a Spotless :)

**Comment by lyharrietbui** [comment]
Hi @nedtwigg I am trying to integrate `diktat` with `spotless`. Does Spotless read the diktat config at configuration-time? It seems like I have to add the `diktat-analysis.yml` at configuration time during the build process for it to be available for 
```
diktat("1.0.3").configFile(diktat-analysis.yml)
```

**Comment by nedtwigg** [comment]
> Does Spotless read the diktat config at configuration-time?

Yes, because it's needed for up-to-date checks. If you want to download a config from a remote URL, I recommend https://github.com/diffplug/blowdryer

**Comment by akuleshov7** [comment]
> > Does Spotless read the diktat config at configuration-time?
> 
> Yes, because it's needed for up-to-date checks. If you want to download a config from a remote URL, I recommend https://github.com/diffplug/blowdryer

@lyharrietbui I do not think that you need to download that from remote URL, you can put it to your project folder. Gradle will automatically calculate paths from the rootDir.

**Pull request merged**